added an undo button

SeanKuehl opened the pull request:
**Fixes/Solves**
Add an undo button. issue #29

I have added an undo button that will restore a previous save state. Right now save states happen in three events: piece added, piece removed and table flipped. These events are also on timers, so after they create a save state they will have to wait out several calls before they can create a save state again.

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -42,20 +42,79 @@ var _srv_allow_card_stacking = true
+var _srv_undo_stack: Array = []
+var _srv_undo_stack_size_limit: int = 10
+
+var _srv_add_piece_timer: int = 0	#this is set to 0 to start to it will save the first state
+var _srv_add_piece_timer_max: int = 5
+
+var _srv_events_add_states: bool = true	#this is so events that normally add states like add_piece don't add a state when they are called as a part of set_state
+
+var _srv_flip_table_timer: int = 0
+var _srv_flip_table_timer_max: int = 5
+
+var _srv_remove_piece_timer: int = 0
+var _srv_remove_piece_timer_max: int = 5
```
A couple of things here:
- Personally, I think instead of having an undo state being added e.g. every 5 pieces, I think there should be a 'cooldown' whenever a piece is added. For example, someone creates a piece, which sets a cooldown for 10 seconds. If someone adds another piece in that 10 seconds, the cooldown is reset back to 10 seconds. Only after the 10 seconds is an undo state created when a piece is added. Let me know what you think about this idea!
- I think instead of having variables for each event, have a dictionary where the keys are the events, and the values are the timers. This way, there aren't as many variables, and more events could easily be added.

SeanKuehl replied:
I have implemented the dictionary and it is now time based(rough amounts of time based on physics process "ticks").

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -42,20 +42,79 @@ var _srv_allow_card_stacking = true
+
+#adds an undo state to the undo stack if possible
+#Returns: nothing(void)
+master func push_undo_state() -> void:
```
`push_undo_state` should not be `master` - only the host dictates when an undo state is pushed onto the stack, and plus, a malicious player could spam remote calls to this function as well.

SeanKuehl replied:
When I did not have the function as master I had errors, is there another keyword like master that would work?

drwhut replied:
Instead of trying to call the function remotely, call it locally (i.e. just call `push_undo_state()`), but make sure that only the host calls it with `get_tree().is_network_server()`.

SeanKuehl replied:
Thank you, I have removed the master keywords and added this check instead to the function called by the undo button's signal.

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -68,31 +127,46 @@ remotesync func add_hand(player: int, transform: Transform) -> void:
+
 	_pieces.add_child(piece)
+	
+	#if not waiting for a timer or disabled because set_state wil be called
+	if _srv_events_add_states and _srv_add_piece_timer == 0:
+		rpc_id(1, "push_undo_state")	#make host add an undo state
+		print("this was called")
+		
+		_srv_add_piece_timer = _srv_add_piece_timer_max	#set timer
+		print(_srv_add_piece_timer)
+	else:
+		#waiting on timer
+		_srv_add_piece_timer -= 1
+		print(_srv_add_piece_timer)
```
Should this not come before the piece is added to the scene tree?

SeanKuehl replied:
Moved it to the top of the add_piece function

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -380,24 +454,33 @@ remotesync func flip_table(camera_basis: Basis) -> void:
 	
+	#if not waiting for a timer or disabled because set_state wil be called
+	if _srv_events_add_states and _srv_flip_table_timer == 0:
+		rpc_id(1, "push_undo_state")	#make host add an undo state
```
This function is called by the host on every player's machine, meaning that this `rpc_id` will be called for every player, making as many undo states as there are players.

As alluded to above as well, I don't think non-host players should have the ability to push onto the undo stack either.

Instead you could check to see if the function is being run on the host machine with `get_tree().is_network_server()`, and only if that returns true, run the `push_undo_state` function.

SeanKuehl replied:
I've added get_tree().is_network_server() as the third requirement in the if statements.

SeanKuehl commented on game/Scripts/Game/CameraController.gd:
```diff
@@ -2048,3 +2051,12 @@ func _on_Viewport_size_changed():
+
+func _on_UndoToolButton_pressed():
+	
+	var room = get_tree().get_current_scene().get_node("Room")	#the pop/push functions are a part of the Room scene's script so we need to get the "Room" node in order to use them
+	
+	room.rpc_id(1, "pop_undo_state")	#make the host pop an und state
```
Hey, I set it up to do the signal but when I try to connect the signal in room it says I'm trying to connect a nonexistent signal. How did you get the other signals to work between scripts?

drwhut replied:
Using the editor, you can follow these steps to connect signals:
1. Go to the scene that contains both nodes that you want to connect (in this case, the camera controller and the room).
2. Click on the node that broadcasts the signal (the camera controller), and on the top right, click `Node`, which should show you the signals.
3. Double-click the new signal, and "attatch" it to the room node. This should create a new function in that node's script that is called when the signal is fired.

SeanKuehl replied:
Thank you, I've hooked up the signal and it works.

SeanKuehl commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -1934,26 +2017,36 @@ func _on_container_absorbing_hovered(container: PieceContainer, player_id: int)
+	
+	
+	#if not waiting for a timer or disabled because set_state wil be called
+	if _srv_events_add_states and _srv_remove_piece_timer == 0:
+		rpc_id(1, "push_undo_state")	#make host add an undo state
+		_srv_remove_piece_timer = _srv_remove_piece_timer_max	#set timer
+	else:
+		#waiting on timer
+		_srv_remove_piece_timer -= 1
```
I've added get_tree().is_network_server() as the third requirement in the if statements.

drwhut commented:
I can see that you've replied to my comments, but I can't see the changes you've made - you'll need to commit and push the changes you've made to your fork's branch.

SeanKuehl commented:
Hey, I have since committed the changes I was able to make. As mentioned in the comments, I have implemented the dictionary system and switched the timers to a 'real time' approach. I also made it so that only the host will save undo states upon events. I do however have some questions about the signal method and the master keyword, hence why these changes have not yet been made.

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -127,6 +129,18 @@ remotesync func add_hand(player: int, transform: Transform) -> void:
 remotesync func add_piece(name: String, transform: Transform,
 	piece_entry: Dictionary) -> void:
 	
+	#if not waiting for a timer or disabled because set_state wil be called
+	if _srv_events_add_states and _srv_undo_state_events["add_piece"] == 0 and get_tree().is_network_server():
+		rpc_id(1, "push_undo_state")	#make host add an undo state
+		
+		print("this was called")
+		
+		_srv_undo_state_events["add_piece"] = _ten_seconds	#set timer
+		
+	else:
+		
+		_srv_undo_state_events["add_piece"] = _ten_seconds	#if you try to add a piece before timer is up, timer resets
+		
```
I would put this after the `if get_tree().get_rpc_sender_id() != 1:` line for security reasons (what is a malicious player spams remote calls to this function?).

SeanKuehl replied:
I have moved the code handling the add_piece event below that if statement

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -1619,6 +1625,21 @@ func _ready():
+			pass
+		#if a timer is != 0, that means it is waiting until it is ready to add a state again
+		else:
+			_srv_undo_state_events[key] -= 1
```
I would use the `delta` value instead of decrementing integers here, just so the timer is accurate.

SeanKuehl replied:
I have made this change. Only note is that it seems that the delta value is much lower than one and the timer length will have to be adjusted. I'll adjust the timer lengths to be reasonable and commit my changes soon.

drwhut commented on game/Scripts/Game/CameraController.gd:
```diff
@@ -2055,8 +2059,11 @@ func _on_Viewport_size_changed():
 
 func _on_UndoToolButton_pressed():
 	
+	#emit_signal("popping_undo_state")
```
This line is correct! It should work once you've connected the signal to the `Room` node via the editor.

SeanKuehl replied:
Yes, now that I have connected the signals it works. Thank you.

drwhut reviewed (changes requested):
Great stuff! The remaining comments are just nit-picky things.

drwhut commented on game/Scripts/Game/3D/Room.gd:
```diff
@@ -45,62 +45,54 @@ var _srv_hand_setup_frames = -1
 #takes an undo state off of the undo stack if possible
 #Returns: nothing(void)
-master func pop_undo_state() -> void:
+func pop_undo_state() -> void:
```
Sorry, I just meant remove the `master` keyword from `push_undo_state()` - any player should be able to undo the game state, but only the host should decide when an undo state is pushed onto the stack.

SeanKuehl replied:
Thanks for clearing that up. I re-added the keyword, changed the call for pop back to rpc_id(1, "pop_undo_state") but removed the code that only let the host do it. Now anyone pressing the undo button can undo, but only the host adds states.

drwhut reviewed (changes requested):
Awesome! One last thing regarding the `master` keyword in `pop_undo_state()`, have a look at the comment.

drwhut commented:
Great! I'll test the branch further later today before I merge it into master. :smile:

drwhut commented:
I've been trying out the branch, and I've come across a potential issue (not a bug per se, but rather a design issue): If you repeat flipping the table, then undoing, quickly enough then there comes a point where you cannot undo, and you cannot get the table back without pressing the "reset table" button. This is due to the fact that the second undo state isn't created because of the 10-second rule.

The reason we introduced the 10-second rule was because of the possibility of players spamming new objects into the game, with each one creating a new undo state - however, I think every table flip should create a new undo state, purely because it is such a 'major' event compared to adding an object.

Maybe we could have it so that there's a constant dictionary of 'timings' for each event, where adding a piece is 10 seconds (as before), removing a piece might be something like 5 seconds, and flipping the table would effectively be 0 seconds. What do you think?

SeanKuehl commented:
Make another dictionary with the same keys "add_piece" but instead the value is what the timer is set to? So for add_piece it would be 10, but then for flip_table it could be 0, I can do that, it should take very long since I'm just replacing the number the timer is set with with the value in a dictionary. I'll also make the timing changes you made in the post.

drwhut commented:
Sorry for the late response, I was gonna check the new commit yesterday, but I ended up getting my first COVID vaccine (yay!) and it wiped me out for the rest of the day.

Thank you so much for contributing to the project! :smile:
